### ISSUE TITLE
Syntax for specifying and ascribing implicit arguments

### ISSUE THREAD

**Issue by brendanzab** [gluon-lang/gluon#98]
At the moment we have one form of implicit argument - type variables. For example:

``` elm
let compose : cat b c -> cat a b -> cat a c = ...
```

Has the implicit arguments:
- `cat : Type -> Type`
- `a : Type`
- `b : Type`

At the moment there is no way to explicitly specify the types of those arguments, or to explicitly ascribe them when mentioning the binding.

As an example, Idris has the following syntax for it:

``` idris
compose : {cat : Type -> Type} -> {b : Type} -> cat b c -> cat a b -> cat a c

foo = compose {a = Int} {cat = (->)}
```

Note in particular that:
- For better or for worse (ask @puffnfresh), variables in lower case are assumed to be implicit args of `Type` - the explicit definition of `b` is just for illustration.
- The brace syntax conflicts with our record syntax.
- This also requires later arguments to depend on previous arguments - perhaps only types would be allowed initially. I think this feature also subsumes [rank-n types](https://wiki.haskell.org/Rank-N_types) - not sure though.

Whatever syntax is chosen, this could eventually be used for modular implicits (see #47). But the restrictions to only using types might have to be relaxed somewhat. I think the [1ML paper](https://www.mpi-sws.org/~rossberg/1ml/1ml-extended.pdf) has some ideas on how to restrict it while retaining decidability (see the section, 'Predicativity').

I'm thinking I might have to start working on a spec for our inference rules before we tackle all this stuff though... I tried using a combination of records and implicit arguments to implement modular implicits, but it couldn't infer the module (EDIT: that was in Idris). :/

**Comment by Marwes** [comment]
> ... variables in lower case are assumed to be implicit args of Type ...

Is this in Idris? In Haskell (and gluon atm) variables default to `Type`/`*` if they are unbound after inference.

>  .... I think this feature also subsumes rank-n types - not sure though.

Haskell has `forall` for this so `(forall a. a ->a) -> () == ({ a: Type } -> a -> a) -> ()`?

> Whatever syntax is chosen ...

Just tossing out ideas.

OCaml has `?` for default arguments so `compose : (cat : Type -> Type)? -> (b : Type)? -> cat b c -> cat a b -> cat a c` may have some precedence. 

Or `compose : [at : Type -> Type] -> [b : Type] -> cat b c -> cat a b -> cat a c` ? Would conflict with array patterns if those were added though.

If we want to make parsing completely insane we could after adding Haskell style declarations allow
```
let compose {cat : Type -> Type} -> {b : Type} : cat b c -> cat a b -> cat a c
    compose ... = ...
```
Only works for types though so maybe just ignore that one...

>  I tried using a combination of records and implicit arguments to implement modular implicits, but it couldn't infer the module. :/

What did you try doing this in?

**Comment by Marwes** [comment]
>The brace syntax conflicts with our record syntax.

Yeah :/. Changing the record syntax would probably be 10x more confusing overall though compared to any other syntax for implicit arguments.

**Comment by brendanzab** [comment]
> What did you try doing this in?

Woops - forgot to say that was in Idris. I was using records for that. I can pop up the code after work.

> Yeah :/. Changing the record syntax would probably be 10x more confusing overall though compared to any other syntax for implicit arguments.

Yeah, we use records as a major syntactic construct - wouldn't want to lose those.

**Comment by brendanzab** [comment]
Lean seems to use `{a : Type}` too, but it isn't as focused on value-level programming.

More 🚲 🏠 :

```idris
compose : ?(cat : Type -> Type, a : Type, b : Type) -> cat b c -> cat a b -> cat a c

compose ?(cat = (->), b = Int) (*) (+)
```

Notice the use of the tuple syntax for more concision, and the partial ascription of `compose`.

**Comment by brendanzab** [comment]
More 🚲 🏠:

- `?{..}` denotes _inferable type_ parameters
  - `?{a}`: kind inferred base on context
  - `?{a b c}`: multiple args with implicit kinds
  - `?{a : Type}` : explicit kind
  - `?{a b c : Type}` : multiple args with explicit kinds
- `?[..]` denotes _implicit value_ parameters
  - `?[X]` : anonymous implicit value
  - `?[x : X]` : named implicit value
  - `?[x y z : X]` : multiple named implicit values

Example:

```elm
type Eq a = {
    eq : a -> a -> Bool,
}

let eq : ?{a} -> ?[e : Eq a] -> a -> a =
    \x y -> e.eq x y

let eq_int : Eq Int = {
    eq = ..,
}

eq 1 2
eq ?{a = Int} 1 2
eq ?[e = eq_int] 1 2
eq ?{a = Int} ?[e = eq_int] 1 2
```

Not sure about the noisiness of `?` though. We _could_ instead add a sigil to records instead, or require a `record` keyword:

```elm
type Eq a = record {
    eq : a -> a -> Bool,
}

let eq : {a} -> [e : Eq a] -> a -> a =
    \x y -> e.eq x y

let eq_int : Eq Int = record {
    eq = ..,
}

eq 1 2
eq {a = Int} 1 2
eq [e = eq_int] 1 2
eq {a = Int} [e = eq_int] 1 2
```

We might need to have a way of explicitly registering specific types and values with the implicit elaborator to reduce the search space. Not sure though 🤔

**Comment by Marwes** [comment]
> We might need to have a way of explicitly registering specific types and values with the implicit elaborator to reduce the search space. Not sure though 🤔

So this is what I mostly think about now. I think it is required that we have some way to limit what can be selected as an implicit argument. This could either be done as something marked per *type* (say `implicit type Test = ...`) or per binding (`implicit let test_instance`),

Embedding it into the type itself has the advantage that we can just search all bindings in scope (including the fields of any records) and we easily have all potential candidates for an arguement (this can be cached so it should not be very expensive). This also gives behaviour very close to normal type classes.

Marking value bindings as implicit is more flexible  but it comes with the problem of *propagating* implicitness.

```f#
// module gets the type `{ x : Int }`
let module =
    implicit let x = 123
    { x }

// How do we determine that `module.x` can be used as the implicit argument?
let f : [y : Int] -> Int = ...

```

Best idea I got so far is that `module` gets the type `{ implicit x : Int }` and then we can search (just as in the first instance) for any bindings marked as implicit... However you then get problems such as, do `{ x : Int } <=> { implicit x : Int }` unify?

**Comment by Marwes** [comment]
> Not sure about the noisiness of ? though. We could instead opt to disambiguate records:

I'd rather have some noisiness when specifying implicit arguments explicitly than the other way around. Specifying implicit arguments explicitly should be fairly rare

**Comment by brendanzab** [comment]
> I'd rather have some noisiness when specifying implicit arguments explicitly than the other way around. Specifying implicit arguments explicitly should be fairly rare

Yeah, we can see how the syntax looks over time based on use, and decide from there.

**Comment by brendanzab** [comment]
More syntax 🚲 🏠:

Seen Conor McBride use this syntax:

```
let x : {?a : Type} -> [?eq : Eq a] -> a -> a -> Bool =
    \x y -> eq.(==) x y

x {?a = Int} [?eq = eq_Int] 1 2
```

Somehow seems cleaner than:

```
let x : ?{a : Type} -> ?[eq : Eq a] -> a -> a -> Bool =
    \x y -> eq.(==) x y

x ?{a = Int} ?[eq = eq_Int] 1 2
```
... I dunno 🤷‍♂️

**Comment by Marwes** [comment]
Looks cleaner for one argument I guess but does it become `{?a, ?b}` or `{?a, b}`? :thinking:

**Comment by brendanzab** [comment]
Bleh, true that.

**Comment by Marwes** [comment]
🚲 🏠

```f#
Eq a => a -> a -> Bool
```

## Pros

* No collision with array or record syntax
* Looks like class constraints in Haskell

## Cons

* No idea how to bring implicit type parameters

**Comment by brendanzab** [comment]
> No idea how to bring implicit type parameters

Could you elaborate on this?

**Comment by Marwes** [comment]
> Could you elaborate on this?

@brendanzab  updated

**Comment by Marwes** [comment]
For reference, the current syntax can be found here https://github.com/gluon-lang/gluon/blob/master/TUTORIAL.md#implicit-arguments
